Starting documentation for Operator version 6 (#2141)
minio/operator: docs/notes/v6.0.0.md

# MinIO Operator v6.0.0 release notes

## About this release

> MinIO Operator is the production-ready easiest way to manage MinIO deployments in kubernetes. And with a new version, new features, a lot of bug fixes, new kubernetes base version requirements, deprecations and much more.

Reviewer: ```suggestion
MinIO Operator is the production-ready easiest way to manage MinIO deployments in kubernetes and with a new version, new features, a lot of bug fixes, new kubernetes base version requirements, deprecations and much more.
```

Reviewer 2: @shtripat fyi I updated your review suggestion with some additional edits

## What’s new?

> Operator 6.0.0 is now available, this release requires Kubernetes version 1.25.0 or later, you must upgrade your Kubernetes cluster to 1.25.0 or later to use Operator v6.0.0+.

Reviewer 2: A little reword, if you like
```suggestion
Operator 6.0.0 is now available. This release requires Kubernetes version 1.25.0 or later. If your cluster does not use Kubernetes version 1.25.0 or later, you must upgrade your cluster to before installing or upgrading to Operator v6.0.0+.
```

* Introducing MinIO Job
* Sidecar container
* TLS refactors
* Operator STS
* Fields immutable
* Field `spec.pools.*.name` is required starting v5.0.15
* Headless service port name is now renamed based on TLS settings

## Introducing MinIO Job

> MinIO Job is designed to manage MinIO using Jobs. Typically, the DevOps team must perform tasks on the fly while deploying MinIO, such as creating a bucket, adding a policy, attaching the policy to a user, and so on. With the current approach, we either continuously modify our existing Tenant CRD, which is already overpopulated, or we adopt a new approach to distribute the load between what a Tenant is supposed to do and what a new controller can handle if the MinIO Job CRD is added. In other words, MinIO Job allows you to manage some of these tasks imperatively through a new CRD. It is like controlling things with mc but via CRD. Please review the examples below for a better understanding of this added feature.

Reviewer 2: Here's an alternative that breaks up the long paragraph. Is this still an accurate summary of the feature? 
```suggestion
MinIO Job is designed to manage MinIO using Jobs. Typically, the DevOps team must perform tasks on the fly while deploying MinIO, such as creating a bucket, adding a policy, attaching the policy to a user, and so on.

The current approach requires continuously modifying the existing Tenant CRD, causing it to become more difficult to manage. The new approach distributes these tasks between the Tenant and a new controller by creating an additional MinIO Job CRD.

MinIO Job allows you to manage tasks imperatively through this additional CRD, reducing complexity. It is like controlling things with mc but using a CRD instead of individual commands.

Please review the examples below for a better understanding of this added feature.
```

### MinIO Job Example (To create a bucket with a Job):

> Assuming that Operator is installed and latest version is used

Deploy `MinIOJob` CRD along with its binded policy:

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: mc-job-sa
  namespace: minio-tenant-1
---
apiVersion: sts.min.io/v1alpha1
kind: PolicyBinding
metadata:

>   name: mc-job-bingding

Reviewer: ```suggestion
  name: mc-job-binding
```

  namespace: minio-tenant-1
spec:
  application:
    serviceaccount: mc-job-sa
    namespace: minio-tenant-1
  policies:
    - consoleAdmin
---
apiVersion: job.min.io/v1alpha1
kind: MinIOJob
metadata:
  name: minio-test-job
  namespace: minio-tenant-1
spec:
  serviceAccountName: mc-job-sa
  tenant:
    name: myminio
    namespace: minio-tenant-1
  commands:
    - op: make-bucket
      args:
        name: memes
```

> 3. Job is completed, bucket created:

Reviewer: What is `3. ` bullet here?

Author: Just to show it got created, but we can remove it.

![bucket creation](images/v6.0.0/image1.png )

### Sidecar container

TODO

### TLS Refactor

TODO

### Operator STS 

API `sts.min.io` is updated from `v1alpha1` to `v1beta1`.

The [PolicyBinding](https://github.com/minio/operator/blob/master/docs/policybinding_crd.adoc#k8s-api-github-com-minio-operator-pkg-apis-sts-min-io-v1alpha1-policybinding)
CRD will be backward compatible.
If a resource of type `PolicyBindig` on version `v1alpha` currently exists in the Cluster
at the moment of the upgrade of Operator to Version 6.0.0, the resource will be automatically updated to version `v1beta1`
with no manual intervention.

<details>
  <summary>Example</summary>

Before update:
```yaml
apiVersion: sts.min.io/v1alpha
kind: PolicyBinding
metadata:
  name: binding-1
  namespace: minio-tenant-1
spec:
  application:
    namespace: sts-client
    serviceaccount: stsclient-sa
  policies:
    - test-bucket-rw
```

After update:
```yaml
apiVersion: sts.min.io/v1beta1
kind: PolicyBinding
metadata:
  name: binding-1
  namespace: minio-tenant-1
spec:
  application:
    namespace: sts-client
    serviceaccount: stsclient-sa
  policies:
    - test-bucket-rw
```
</details>

### Fields immutable

> `spec.pools.*.volumesPerServer` and `spec.pools.*.servers` fields are immutable once created, this was introduced in 
> https://github.com/minio/operator/pull/2070 on Operator v5.0.15, this is a safety net to prevent potentially dangerous change,
> once a pool is created, and it contains data the pool's number of servers and the number of volumes per server should not be changed.

Reviewer 2: ```suggestion
`spec.pools.*.volumesPerServer` and `spec.pools.*.servers` fields are immutable once created, ensuring the number of servers or volumes in an existing pool does not change. It was introduced in Operator v5.0.15 with PR https://github.com/minio/operator/pull/2070 and is a safety net to prevent potentially dangerous changes to an existing pool. This also matches similar behavior in Operator Console.
```

> Some of the reasons might be in need to modify this values are expanding storage, this should be made adding a pool,
> see [Expand a distributed MinIO Deployment](https://min.io/docs/minio/linux/operations/install-deploy-manage/expand-minio-deployment.html)
> for this.
> 
> On the other hand, the removal of servers no longer in use or ready to be decommissioned should be carefully taken care 
> of by [Decommission a Tenant Server Poo](https://min.io/docs/minio/kubernetes/upstream/operations/install-deploy-manage/modify-minio-tenant.html#decommission-a-tenant-server-pool) 
> not individual nodes or drives.

Reviewer 2: ```suggestion
To expand storage, add an additional pool.
See [Expand a distributed MinIO Deployment](https://min.io/docs/minio/linux/operations/install-deploy-manage/expand-minio-deployment.html)
for more information.

To reduce storage capacity, remove pools no longer in use by decommissioning them. See [Decommission a Tenant Server Pool](https://min.io/docs/minio/kubernetes/upstream/operations/install-deploy-manage/modify-minio-tenant.html#decommission-a-tenant-server-pool) 
for details.
```

If a drive is misbehaving or failing follow the [Drive Failure Recovery](https://min.io/docs/minio/kubernetes/upstream/operations/data-recovery/recover-after-drive-failure.html#minio-restore-hardware-failure-drive)
guide, if a node is failing follow Node [Failure Recovery guide](https://min.io/docs/minio/kubernetes/upstream/operations/data-recovery/recover-after-node-failure.html).

### Field `spec.pools.*.name` is required starting v5.0.15

TODO

### Headless service port name is now renamed based on TLS settings

Headless service now changes port name from `http-minio` to `https-minio` when TLS is enabled, be aware to configure Ingres
or any other resources that reference the port name accordingly.

When `.spec.requestAutocert` is `false` and `.spec.externalCertSecret` is empty, then no TLS encryption in transit 

> (https) is enabled for the tenant, and the 3 services created in the namespace have a prefixed `http-` service port:

Reviewer: ```suggestion
(https) is enabled for the tenant, and the 3 services created in the namespace have a prefix `http-` service port:
```

# MinIO Operator v6.0.0 release notes

## About this release

MinIO Operator is the recommended production-ready way to manage MinIO deployments in Kubernetes. This version includes many new features, bug fixes, new Kubernetes base version requirements, deprecations, and more.

## What’s new?

Operator 6.0.0 is now available. This release requires Kubernetes version 1.25.0 or later. If your cluster does not use Kubernetes version 1.25.0 or later, you must upgrade your cluster to before installing or upgrading to Operator v6.0.0+.

* Introducing MinIO Job
* Sidecar container
* TLS refactors
* Operator STS

> * Fields immutable

Reviewer 2: This might make more sense to somebody who doesn't read our Operator repo issues all the time 😄 
```suggestion
* `servers` and `volumesPerServer` fields are immutable
```

        name: memes
```

![bucket creation](images/v6.0.0/image1.png )

### Sidecar container

TODO

### TLS Refactor

TODO

### Operator STS 

API `sts.min.io` is updated from `v1alpha1` to `v1beta1`.

The [PolicyBinding](https://github.com/minio/operator/blob/master/docs/policybinding_crd.adoc#k8s-api-github-com-minio-operator-pkg-apis-sts-min-io-v1alpha1-policybinding)
CRD will be backward compatible.
If a resource of type `PolicyBinding` on version `v1alpha` currently exists in the Cluster
at the moment of the upgrade of Operator to Version 6.0.0, the resource will be automatically updated to version `v1beta1`
with no manual intervention.

<details>
  <summary>Example</summary>

Before update:
```yaml
apiVersion: sts.min.io/v1alpha
kind: PolicyBinding
metadata:
  name: binding-1
  namespace: minio-tenant-1
spec:
  application:
    namespace: sts-client
    serviceaccount: stsclient-sa
  policies:
    - test-bucket-rw
```

After update:
```yaml
apiVersion: sts.min.io/v1beta1
kind: PolicyBinding
metadata:
  name: binding-1
  namespace: minio-tenant-1
spec:
  application:
    namespace: sts-client
    serviceaccount: stsclient-sa
  policies:
    - test-bucket-rw
```
</details>

> ### Fields immutable

Reviewer 2: ```suggestion
### `servers` and `volumesPerServer` fields are immutable
```

`spec.pools.*.volumesPerServer` and `spec.pools.*.servers` fields are immutable once created, ensuring the number of servers or volumes in an existing pool does not change. It was introduced in Operator v5.0.15 with PR https://github.com/minio/operator/pull/2070 and is a safety net to prevent potentially dangerous changes to an existing pool. This also matches similar behavior in Operator Console.

To expand storage, add an additional pool.
See [Expand a distributed MinIO Deployment](https://min.io/docs/minio/linux/operations/install-deploy-manage/expand-minio-deployment.html)
for more information.

To reduce storage capacity, remove pools no longer in use by decommissioning them. See [Decommission a Tenant Server Pool](https://min.io/docs/minio/kubernetes/upstream/operations/install-deploy-manage/modify-minio-tenant.html#decommission-a-tenant-server-pool) 
for details.

> If a drive is misbehaving or failing follow the [Drive Failure Recovery](https://min.io/docs/minio/kubernetes/upstream/operations/data-recovery/recover-after-drive-failure.html#minio-restore-hardware-failure-drive)
> guide, if a node is failing follow Node [Failure Recovery guide](https://min.io/docs/minio/kubernetes/upstream/operations/data-recovery/recover-after-node-failure.html).

Reviewer 2: ```suggestion
If a drive is misbehaving or failing, follow the [Drive Failure Recovery](https://min.io/docs/minio/kubernetes/upstream/operations/data-recovery/recover-after-drive-failure.html#minio-restore-hardware-failure-drive)
guide. If a node is failing, follow the [Node Failure Recovery guide](https://min.io/docs/minio/kubernetes/upstream/operations/data-recovery/recover-after-node-failure.html).
```

### Field `spec.pools.*.name` is required starting v5.0.15

TODO

### Headless service port name is now renamed based on TLS settings

Headless service now changes port name from `http-minio` to `https-minio` when TLS is enabled, be aware to configure Ingres
or any other resources that reference the port name accordingly.

When `.spec.requestAutocert` is `false` and `.spec.externalCertSecret` is empty, then no TLS encryption in transit 
(https) is enabled for the tenant, and the 3 services created in the namespace have a prefix `http-` service port:

`{tenant-name}-console` exposes port `http-console`

<details>
  <summary> example</summary>

```yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    v1.min.io/console: tenant-1-console
  name: tenant-1-console
  namespace: tenant-1
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 10.96.78.111
  clusterIPs:
  - 10.96.78.111
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  ipFamilyPolicy: SingleStack
  ports:
  - name: http-console
    nodePort: 30984
    port: 9090
  selector:
    v1.min.io/tenant: tenant-1
  type: LoadBalancer
```
</details>

> the same, `minio` exposes port `http-minio` and headless service `{tenant-name}-hl` exposes port `http-minio`.

Reviewer 2: I'm not sure, but I think might be like this wording?
```suggestion
Similarly, `minio` exposes port `http-minio` and headless service `{tenant-name}-hl` exposes port `http-minio`.
```

# MinIO Operator v6.0.0 release notes

## About this release

MinIO Operator is the production-ready easiest way to manage MinIO deployments in kubernetes. And with a new version, new features, a lot of bug fixes, new kubernetes base version requirements, deprecations and much more.

## What’s new?

> Operator 6.0.0 is now available, this release requires Kubernetes version 1.25.0 or later, you must upgrade your Kubernetes cluster to 1.25.0 or later to use Operator v6.0.0+.

Reviewer 3: @cniackz  @dvaldivia why is needed to start on Kubernetes 1.25.0 and what happens with installs on Kubernetes 1.21 or previous? Are these installs no longer supported?

Author: We have been encouraging users to adhere to the supported Kubernetes versions listed on the [Kubernetes releases page](https://kubernetes.io/releases/). Version 1.21 is no longer supported, so we need to upgrade to a newer, supported version based on the `End of Life` schedule.

* Introducing MinIO Job

> * Sidecard container

Reviewer: ```suggestion
* Sidecar container
```

* TLS refactors
* Operator STS

> * Fields immutable

Reviewer 3: ```suggestion
## Fields immutable

`spec.pools.*.volumesPerServer` and `spec.pools.*.servers` fields are immutable once created, this was introduced in https://github.com/minio/operator/pull/2070 on v5.0.15
```

* Field `spec.pools.*.name` is required starting v5.0.15
* Headless service port name is now renamed based on TLS settings

## Introducing MinIO Job

MinIO Job is designed to manage MinIO using Jobs. Typically, the DevOps team must perform tasks on the fly while deploying MinIO, such as creating a bucket, adding a policy, attaching the policy to a user, and so on. With the current approach, we either continuously modify our existing Tenant CRD, which is already overpopulated, or we adopt a new approach to distribute the load between what a Tenant is supposed to do and what a new controller can handle if the MinIO Job CRD is added. In other words, MinIO Job allows you to manage some of these tasks imperatively through a new CRD. It is like controlling things with mc but via CRD. Please review the examples below for a better understanding of this added feature.

### MinIO Job Example (To create a bucket with a Job):

> Assuming that Operator is installed and latest version is used

> 1. Enable STS:
> 
> ```shell
> kubectl -n minio-operator set env deployment/minio-operator OPERATOR_STS_ENABLED=on
> ```
> 
> 2. Deploy `MinIOJob` CRD along with its binded policy:

Reviewer 3: ```suggestion
Deploy `MinIOJob` CRD along with its binded policy:
```

Operator STS is already enabled by default, even if no env variable `OPERATOR_STS_ENABLED` is declared
